### PR TITLE
feat(surveys): adds targeting changes to the survey revision history

### LINKED ISSUE #23725: Support survey targeting revision history
## Is your feature request related to a problem?

In this change: https://github.com/PostHog/posthog/pull/23622, we added revision history logging for a subset of survey change operations (details in that PR), but we didn't cover changes made to targeting logic that leverages feature flags under the hood.  It's a bit more complex for the following reasons:

- `changes_between` doesn't handle the `remove_targeting_flag` case well since it relies on reading a Survey model that has all of its relationships intact in order to calculate values, but that fails whenever a targeting flag is deleted because deleting a targeting flag deletes the underlying feature flag, and therefore makes it impossible to look up via `getattr` in `changes_between`
- we haven't written any serializers to handle parsing the feature flag changes in the context of a survey revision

## Describe the solution you'd like

I think we should do the following

- make `changes_between` work when a foreignkey relationship is deleted between `previous` and `current`
- add `FeatureFlag` parsing support to the `Changes` object so we can visualize it on the UI.

### PR DESCRIPTION
**Pull request by dmarticus**
Closes https://github.com/PostHog/posthog/issues/23725

A few interesting things that I did (things worth looking at in greater depth IMO)

- Modified `surveys.py` to handle manually inserting changelog entries whenever we make changes to the survey targeting (this has to be done manually because the existing `changes_between` abstraction only supports calculating changes within an existing model, and doesn't support foreign key relationships.  I wrote a test for this new behavior.
- Put way too much work into `surveyActivityDescriber` to handle displaying changes to survey components in a easy-to-digest way.  Outside of just writing a lot of code, these changes included:
    - introduced [ts-pattern](https://github.com/gvergnaud/ts-pattern) to do structured pattern matching of tuples so that I didn't have to write infinite chains of conditionals when pattern matching changes to question types.  This dependency is [quite small](https://bundlephobia.com/package/ts-pattern@3.0.2), and I think the developer ergonomics of introducing structural pattern matching to our codebase outweigh the slightly bigger bundle.
    - a bunch of unit tests for the various different display functions in this method, so it's easier to change it as we continue to go along.
- Added support for parsing FeatureFlag models in `activity_log`
- Added error handling to the `changes_between` method to prevent it from failing when it tried to look up objects that didn't exist.  That's by design!

## Changes

Feel free to pull this to your local branch and add some targeting behavior yourself, here's what my survey history looks like after a bunch of manual testing

## Does this work well for both Cloud and self-hosted?

Yes

## How did you test this code?

- manual tests for UI
- unit tests for the server-side stuff
- unit tests for the various information-distilling sub-methods in `surveyActivityDescriber`

### CODE REVIEW

File: frontend/src/scenes/feature-flags/activityDescriptions.tsx [hunk 1]
```diff
@@ -260,6 +260,22 @@ export function flagActivityDescriber(logItem: ActivityLogItem, asNotification?:
         return { description: null }
     }
 
+    if (logItem.activity === 'created') {
```
Comment (dmarticus) [inline frontend/src/scenes/feature-flags/activityDescriptions.tsx:263]: drive by PR when I noticed that we never actually visualize (but we _do_ log) creation events in the feature flag history!  Here's what it look like now

image

File: frontend/src/scenes/surveys/surveyActivityDescriber.tsx [hunk 2]
```diff
@@ -31,28 +42,102 @@ const surveyActionsMapping: Record<
...
+            description: [
+                <>
+                    updated <strong>{questionChanges.length}</strong> question{questionChanges.length !== 1 ? 's' : ''}:
+                    <ul className="bullet-list">
+                        {questionChanges.map(({ index, changes }) => (
+                            <li key={index}>
+                                Question {index}:
+                                <ul className="bullet-list">
+                                    {changes.map((changeItem, changeIndex) => (
+                                        <li key={changeIndex}>{changeItem}</li>
```
Comment (dmarticus) [inline frontend/src/scenes/surveys/surveyActivityDescriber.tsx:133]: this gives us a UI like this, which I'm pretty happy with (although the "Question 1" bit seems bit clunky, I just wanted to let them know which question was actually changed!)

image

File: frontend/src/scenes/surveys/surveyActivityDescriber.tsx [hunk 3]
```diff
@@ -149,161 +211,125 @@ const surveyActionsMapping: Record<
...
+                beforeConditions?.seenSurveyWaitPeriodInDays,
+                afterConditions?.seenSurveyWaitPeriodInDays,
+                'days'
+            ),
+            describeFieldChange('URL match type', beforeConditions?.urlMatchType, afterConditions?.urlMatchType)
+        )
+
+        // Use JSON.stringify for deep comparison of objects
+        if (JSON.stringify(beforeConditions?.events) !== JSON.stringify(afterConditions?.events)) {
+            changes.push(<>modified event conditions</>)
```
Comment (dmarticus) [inline frontend/src/scenes/surveys/surveyActivityDescriber.tsx:228]: elected to not go into details about the event condition changes, just letting users know that they were changed at all.

File: frontend/src/scenes/surveys/surveyActivityDescriber.tsx [hunk 4]
```diff
@@ -410,3 +440,177 @@ function getPreposition(field: string): string {
...
+    const typeChangeDescription =
+        before.type !== after.type
+            ? [
+                  <>
+                      changed question type from <strong>{before.type}</strong> to <strong>{after.type}</strong>
+                  </>,
+              ]
+            : []
+
+    const specificChanges = match([before, after] as const)
```
Comment (dmarticus) [inline frontend/src/scenes/surveys/surveyActivityDescriber.tsx:457]: some cool shit in here with `ts-pattern` that lets me match a tuple of types with a very nice combinator-style developer experience.

File: frontend/src/scenes/surveys/surveyActivityDescriber.tsx [hunk 5]
```diff
@@ -410,3 +440,177 @@ function getPreposition(field: string): string {
...
+                { type: P.union(SurveyQuestionType.SingleChoice, SurveyQuestionType.MultipleChoice) },
+            ],
+            describeMultipleChoiceChanges
+        )
+        .otherwise(() => [])
+
+    return [...commonChanges, ...typeChangeDescription, ...specificChanges, ...describeBranchingChanges(before, after)]
+}
+
+export function describeCommonChanges(before: SurveyQuestion, after: SurveyQuestion): JSX.Element[] {
```
Comment (dmarticus) [inline frontend/src/scenes/surveys/surveyActivityDescriber.tsx:472]: I still have functions that are comparing outcomes conditionally, but now that they can be flattened by the `.with()` combinator, this whole chunk of code gets easier to read.

File: frontend/src/scenes/surveys/surveyActivityDescriber.test.tsx [hunk 6]
```diff
@@ -0,0 +1,260 @@
...
+        buttonText: 'Continue',
+    }
+
+    test('describes common changes', () => {
+        const changes = describeCommonChanges(before, after)
+        expect(changes).toHaveLength(4)
+        expect(getTextContent(changes[0])).toBe(
+            'changed question text from "What is your favorite color?" to "What is your favorite animal?"'
+        )
+        expect(getTextContent(changes[1])).toBe('updated description')
```
Comment (jurajmajerik) [inline frontend/src/scenes/surveys/surveyActivityDescriber.test.tsx:127]: why not also include the before/after description, like with the question change above?

File: frontend/src/scenes/surveys/surveyActivityDescriber.test.tsx [hunk 7]
```diff
@@ -0,0 +1,260 @@
...
+        lowerBoundLabel: 'Bad',
+        upperBoundLabel: 'Good',
+    }
+
+    test('describes rating changes', () => {
+        const changes = describeRatingChanges([before, after])
+        expect(changes).toHaveLength(3)
+        expect(getTextContent(changes[0])).toBe('changed rating display from emoji to number')
+        expect(getTextContent(changes[1])).toBe('changed rating scale from 5 to 10')
+        expect(getTextContent(changes[2])).toBe('updated rating labels from Poor-Excellent to Bad-Good')
```
Comment (jurajmajerik) [inline frontend/src/scenes/surveys/surveyActivityDescriber.test.tsx:173]: Multi-word labels will be rendered like this `...from Very bad-Very good to Quite bad-Quite good`

I'd suggest double quotes here: `...from "Very bad"-"Very good" to "Quite bad"-"Quite good"`

File: frontend/src/scenes/surveys/surveyActivityDescriber.tsx (hunk 3 above)
Comment (jurajmajerik) [reply frontend/src/scenes/surveys/surveyActivityDescriber.tsx:228]: reasonable for now 👍

**Review by jurajmajerik — APPROVED** [review]
fantastic work 🙌 a few minor comments

File: frontend/src/scenes/surveys/surveyActivityDescriber.test.tsx (hunk 6 above)
Comment (dmarticus) [reply frontend/src/scenes/surveys/surveyActivityDescriber.test.tsx:127]: overlooked it; great feedback. Will add.

**Pull request merged**